### PR TITLE
[#215] Add custom options loader at runtime for Kafka API and upgrade kafka clients api to 2.2.1

### LINKED ISSUE #215: [Enhancement] Add custom options loader at runtime for Kafka API and upgrade kafka clients api to 2.2.1
**What feature are your requesting?**
- Currently, by default bootstrap servers are passed to kafka consumer.
But along with that if a user is interested to load the kafka properties from an external library or service, this can be done by plugging the loader class at runtime (which should extend a common interface).
- Upgrade kafka-clients api to 2.2.1

**What benefits do you foresee from the feature you are requesting?**
Increasing the extensibility of Gimel Kafka API

**Potential solution/ideas?**
Add an interface in gimel-common for Options Loader so that anyone can extend it to create their custom loader.

**Additional context**

### PR DESCRIPTION
**Pull request by anishanainani**
Make sure you have checked all steps below.

### GitHub Issue
Fixes #215 

### Checklist:

- [x] This pull request updates the documentation
- [x] This pull request changes library dependencies
- [x] Title of the PR is of format (example) : [#25][Github] Add Pull Request Template

### What is the purpose of this pull request?
- Currently, by default bootstrap servers are passed to kafka consumer.
But along with that if a user is interested to load the kafka properties from an external library or service, this can be done by plugging the loader class at runtime (which should extend a common interface).
- Upgrade kafka-clients api to 2.2.1
- Change zookeeper client from twitter to curator 

### How was this change validated?
Spark shell

```
import org.apache.spark.sql.{Column, Row, SparkSession,DataFrame}
import org.apache.spark.sql.functions._
// Create Gimel SQL reference
val gsql: (String) => DataFrame = com.paypal.gimel.sql.GimelQueryProcessor.executeBatch(_: String, spark)
gsql("set gimel.logging.level=CONSOLE")

//Setting UDC rest api properties
gsql("set rest.service.method=https")
gsql("set rest.service.host=udc-rest-api-host")
gsql("set rest.service.port=443")

// Setting kafka properties
gsql("set gimel.kafka.throttle.batch.fetchRowsOnFirstRun=1")
gsql("set gimel.deserializer.class=com.paypal.gimel.deserializers.generic.JsonDynamicDeserializer")
gsql("set gimel.kafka.options.loader=com.test.CustomKafkaOptionsLoader")

// Reading from kafka dataset
val df = gsql("select * from udc.Kafka.Gimel_Dev.default.user")
```

### Commit Guidelines
- [x] My commits all reference GH issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gimel-dataapi/gimel-common/src/main/scala/com/paypal/gimel/common/conf/GimelTypeRefConstants.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+
+package com.paypal.gimel.common.conf
+
+import com.fasterxml.jackson.core.`type`.TypeReference
+
+import com.paypal.gimel.common.gimelservices.payload.Zone
+
+object GimelTypeRefConstants {
```
Comment (Dee-Pac) [inline gimel-dataapi/gimel-common/src/main/scala/com/paypal/gimel/common/conf/GimelTypeRefConstants.scala:26]: Can we not just ENUMs & Maps? This introduces additional library dependencies & also a new pattern is being introduced.
https://github.com/paypal/gimel/blob/cc46175493c516445d488d9ebe73603d49f56a1a/gimel-dataapi/gimel-common/src/main/scala/com/paypal/gimel/common/utilities/DataSetUtils.scala#L36

File: gimel-dataapi/gimel-common/src/test/scala/com/paypal/gimel/common/utilities/storageadmin/ZooKeeperAdminClientTest.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+import com.paypal.gimel.common.storageadmin.ZooKeeperAdminClient
+import com.paypal.gimel.common.utilities.zookeeper.ZooKeeperEmbedded
+import com.paypal.gimel.logger.Logger
+
+class ZooKeeperAdminClientTest extends FunSpec with Matchers with BeforeAndAfterAll {
+
+  val DEFAULT_ZK_PORT = 2181
+  var zookeeper: ZooKeeperEmbedded = _
+  val checkpointRoot = "/pcatalog/kafka_consumer/checkpoint"
```
Comment (Dee-Pac) [inline gimel-dataapi/gimel-common/src/test/scala/com/paypal/gimel/common/utilities/storageadmin/ZooKeeperAdminClientTest.scala:32]: Shall we move this to constants?

Comment (anishanainani) [reply gimel-dataapi/gimel-common/src/test/scala/com/paypal/gimel/common/utilities/storageadmin/ZooKeeperAdminClientTest.scala:32]: Sure

File: gimel-dataapi/gimel-common/src/main/scala/com/paypal/gimel/common/conf/GimelTypeRefConstants.scala (hunk 1 above)
Comment (anishanainani) [reply gimel-dataapi/gimel-common/src/main/scala/com/paypal/gimel/common/conf/GimelTypeRefConstants.scala:26]: @Dee-Pac We are using it for scala ObjectMapper which expects the TypeReference.

**Comment by anishanainani** [comment]
@Dee-Pac I do not see any merge conflict.

**Comment by anishanainani** [comment]
@Dee-Pac Resolved the conflicts

**Pull request merged**